compiler/rustc_parse/src/parser/pat.rs

```diff
@@ -938,16 +938,24 @@ impl<'a> Parser<'a> {
                 }
             }
 
-            fields.push(match self.parse_pat_field(lo, attrs) {
-                Ok(field) => field,
-                Err(err) => {
-                    if let Some(mut delayed_err) = delayed_err {
-                        delayed_err.emit();
-                    }
-                    return Err(err);
-                }
-            });
-            ate_comma = self.eat(&token::Comma);
+            let field =
+                self.collect_tokens_trailing_token(attrs, ForceCollect::No, |this, attrs| {
+                    let field = match this.parse_pat_field(lo, attrs) {
+                        Ok(field) => Ok(field),
+                        Err(err) => {
+                            if let Some(mut delayed_err) = delayed_err.take() {
+                                delayed_err.emit();
+                            }
+                            return Err(err);
+                        }
+                    }?;
+                    ate_comma = this.eat(&token::Comma);
+                    // We just ate a comma, so there's no need to use
+                    // `TrailingToken::Comma`
```

Comment:
I think we should migrate to this scheme everywhere and eliminate `TrailingToken` entirely (not in this PR).

Reply:
The main obstacle to doing this will be `parse_seq_to_before_token` - it relies on being able to eat the separating tokens, and performs some non-trivial recovery logic. However, I have no objection to removing `TrailingToken`.